pkg/plugins/backendplugin/grpcplugin/client_v2.go

```diff
@@ -338,3 +347,35 @@ func (c *ClientV2) ConvertObjects(ctx context.Context, req *backend.ConversionRe
 
 	return backend.FromProto().ConversionResponse(protoResp), nil
 }
+
+// handleGrpcStatusError handles gRPC status errors and sets the error source based on the error metadata.
+func handleGrpcStatusError(ctx context.Context, err error) error {
+	st := status.Convert(err)
+	for _, detail := range st.Details() {
+		switch t := detail.(type) {
+		case *errdetails.ErrorInfo:
+			errorSource, ok := t.Metadata[errorSourceMetadataKey]
+			if !ok {
+				break
+			}
+
+			switch errorSource {
+			case string(backend.ErrorSourceDownstream):
+				innerErr := backend.WithErrorSource(ctx, backend.ErrorSourceDownstream)
+				if innerErr != nil {
+					logger.Error("Could not set downstream error source", "error", innerErr)
+				}
+				// TODO decide if underlying public message should be improved
+				return plugins.ErrPluginDownstreamErrorBase.Errorf("%v", err)
+			case string(backend.ErrorSourcePlugin):
+				errorSourceErr := backend.WithErrorSource(ctx, backend.ErrorSourcePlugin)
+				if errorSourceErr != nil {
+					logger.Error("Could not set plugin error source", "error", errorSourceErr)
+				}
+				// TODO decide if underlying public message should be improved
+				return plugins.ErrPluginDownstreamErrorBase.Errorf("%v", err)
```

Comment:
It's not clear to me if using `ErrPluginDownstreamErrorBase` where the error source is `plugin` is intentional or not. In this case, the function returns an error derived from `ErrPluginDownstreamErrorBase` (which is a downstream error) regardless of the error source:

https://github.com/grafana/grafana/blob/99d8025829e896b68f6cd5eb11d4ff448d6c3b30/pkg/plugins/errors.go#L31

Is this function only supposed to mutate the error source in the `context.Context`? If so, I'd change the docstring to make it more clear :)

Reply:
For now this is still what we want to do I think because `plugin` or `plugin downstream` is all still `downstream` from Grafana's perspective (or at least that's how it has been before)

Will update the doc string 👍

Reply:
nit: I think a line comment here is more readable (easy to find) than the comment in the func comment